src/cargo/core/compiler/build_context/target_info.rs

```diff
@@ -672,13 +672,18 @@ impl RustcTargetData {
         let host_info = TargetInfo::new(config, requested_kinds, &rustc, CompileKind::Host)?;
         let mut target_config = HashMap::new();
         let mut target_info = HashMap::new();
-        for kind in requested_kinds {
-            if let CompileKind::Target(target) = *kind {
+        // Note that we cannot just rely on getting kinds for all units as the units aren't built yet (TODO CHECK THIS IS TRUE?)
+        let all_kinds = requested_kinds
+            .iter()
+            .copied()
+            .chain(ws.members().flat_map(|p| p.manifest().kind().into_iter()));
```

Comment:
I think that this will need to iterate over the entire crate graph instead of just `ws.members()` to find requested targets?

Reply:
I've tried thinking about this, but it looks to me like `RustcTargetData` does not already have information enough to actually list the entire crate graph, and this only happens in `BuildContext` which gets called after that? And as `RustcTargetData` also gets used from eg. `cargo fetch`, which AFAIU does itself not have enough information to iterate over the entire crate graph — and it is indeed using the kind data in `RustcTargetData`.

Now, seeing as the structures filled here are AFAIU just caches, it'd be very possible to just fill them in on-demand instead, which should solve the issue. Would that sound good to you?